src/library/scala/math/BigInt.scala

```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
 
   /** Compares this BigInt with the specified BigInt for equality.
    */
-  def equals (that: BigInt): Boolean = compare(that) == 0
+  def equals (that: BigInt): Boolean =
+    if (this.longEncoding) {
+      if (that.longEncoding) this._long == that._long else false
+    } else {
+      if (that.longEncoding) false else this._bigInteger == that._bigInteger
+    }
 
   /** Compares this BigInt with the specified BigInt
    */
-  def compare (that: BigInt): Int = this.bigInteger.compareTo(that.bigInteger)
+  def compare (that: BigInt): Int =
+    if (this.longEncoding) {
+      if (that.longEncoding) java.lang.Long.compare(this._long, that._long) else -that._bigInteger.signum()
+    } else {
+      if (that.longEncoding) this.signum else this._bigInteger.compareTo(that._bigInteger)
+    }
 
   /** Addition of BigInts
    */
-  def +  (that: BigInt): BigInt = new BigInt(this.bigInteger.add(that.bigInteger))
+  def +  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      val x = this._long
+      val y = that._long
+      val z = x + y
+      if ((~(x ^ y) & (x ^ z)) >= 0L) return new BigInt(z)
```

Comment:
can you comment why `new BigInt(z)` here and not `return BigInt(z)` which could take advantage of cached boxed values?

Reply:
The original code, for some reason, did not use the cache for unary or binary operations. I wonder if there was any reason (conditional tests being expensive?). I made two variants, the one using the cache is here: https://github.com/denisrosset/scala/commits/bigint-cached

In the factorial benchmark above, which doesn't hit the cache a lot but doesn't stress the memory either, there is basically no speed difference. I'll be merging the caching changes.